program-runtime/src/loaded_programs.rs

```diff
@@ -818,11 +818,33 @@ impl<FG: ForkGraph> ProgramCache<FG> {
             &entry.program,
             ProgramCacheEntryType::DelayVisibility
         ));
+        // This function always returns `true` during normal operation.
+        // Only during the recompilation phase this can return `false`
+        // for entries with `upcoming_environments`.
+        fn is_current_env(
+            environments: &ProgramRuntimeEnvironments,
+            env_opt: Option<&ProgramRuntimeEnvironment>,
+        ) -> bool {
+            env_opt
+                .map(|env| {
+                    Arc::ptr_eq(env, &environments.program_runtime_v1)
+                        || Arc::ptr_eq(env, &environments.program_runtime_v2)
+                })
+                .unwrap_or(true)
+        }
         let slot_versions = &mut self.entries.entry(key).or_default();
         match slot_versions.binary_search_by(|at| {
             at.effective_slot
                 .cmp(&entry.effective_slot)
                 .then(at.deployment_slot.cmp(&entry.deployment_slot))
+                .then(
+                    // This `.then()` has no effect during normal operation.
+                    // Only during the recompilation phase this does allow entries
+                    // which only differ in their environment to be interleaved in `slot_versions`.
+                    is_current_env(&self.environments, at.program.get_environment()).cmp(
+                        &is_current_env(&self.environments, entry.program.get_environment()),
+                    ),
+                )
```

Comment:
Is there a test that explicitly covers these new lines?

Reply:
If I comment these lines out, none of the tests you mentioned failed. 

Going further, I replaced your lines by `false.cmp(&false)`, `true.cmp(&true)`, `false.cmp(&true)` and `true.cmp(&false)` and got no failures either.

Reply:
Ah, now I get what you were asking. These two test just exercise the code, hence they had to be adjusted.

But the regression test for this change is in: `test_feature_activation_loaded_programs_epoch_transition()` try that one with and without the `.then()`.

Reply:
Those are working. Thanks for the clarification. I didn't see any regression tests in this PR, so I asked.